src/model/differ.js

```diff
@@ -0,0 +1,774 @@
+		this._makeSnapshot( parent );
+
+		// Then, get all changes that already were done on the element (empty array if this is the first change).
+		const changes = this._getChangesForElement( parent );
+
+		// Then, look through all the changes, and transform them or the new change.
+		for ( const oldChange of changes ) {
+			this._handleChange( changeItem, oldChange, changes );
+		}
+
+		// Add the new change.
+		changes.push( changeItem );
+
+		// Remove incorrect changes. During transformation some change might be, for example, included in another.
+		// In that case, the change will have `howMany` property set to `0` or less. We need to remove those changes.
+		for ( let i = 0; i < changes.length; i++ ) {
+			if ( changes[ i ].howMany < 1 ) {
+				changes.splice( i, 1 );
+
+				i--;
+			}
+		}
+	}
+
+	/**
+	 * Gets an array of changes that were already saved for given element.
+	 *
+	 * @private
+	 * @param {module:engine/model/element~Element} element
+	 * @returns {Array.<Object>}
+	 */
+	_getChangesForElement( element ) {
+		let changes;
+
+		if ( this._changesInElement.has( element ) ) {
+			changes = this._changesInElement.get( element );
+		} else {
+			changes = [];
+
+			this._changesInElement.set( element, changes );
+		}
+
+		return changes;
+	}
+
+	/**
+	 * Saves a children snapshot for given element.
+	 *
+	 * @private
+	 * @param {module:engine/model/element~Element} element
+	 */
+	_makeSnapshot( element ) {
+		if ( !this._elementSnapshots.has( element ) ) {
+			this._elementSnapshots.set( element, _getChildrenSnapshot( element.getChildren() ) );
+		}
+	}
+
+	/**
+	 * For given newly saved change, compares it with a change already done on the element and modifies the incoming
+	 * change and/or the old change.
+	 *
+	 * @private
+	 * @param {Object} inc Incoming (new) change.
+	 * @param {Object} old Old change (already done on the element).
+	 * @param {Array.<Object>} changes Array containing all the changes done on that element.
+	 */
+	_handleChange( inc, old, changes ) {
+		const incEnd = inc.offset + inc.howMany;
+		const oldEnd = old.offset + old.howMany;
+
+		if ( inc.type == 'insert' ) {
+			if ( old.type == 'insert' ) {
+				if ( inc.offset <= old.offset ) {
+					old.offset += inc.howMany;
+				} else if ( inc.offset < oldEnd ) {
+					old.howMany += inc.howMany;
+					inc.howMany = 0;
+				}
+			}
+
+			if ( old.type == 'remove' ) {
+				if ( inc.offset < old.offset ) {
+					old.offset += inc.howMany;
+				}
+			}
+
+			if ( old.type == 'attribute' ) {
+				if ( inc.offset <= old.offset ) {
+					old.offset += inc.howMany;
+				} else if ( inc.offset < oldEnd ) {
+					const howMany = old.howMany;
+
+					old.howMany = inc.offset - old.offset;
+
+					// Unshift to prevent further processing of this change.
+					changes.unshift( { type: 'attribute', offset: incEnd, howMany: howMany - old.howMany, count: this._changeCount++ } );
```

Comment:
Comment why you add change in the first place is needed.

Reply:
And since you unshift because you know that this function is called in the loop, the loop should be moved inside the function.